tool-cache extractZip not working

I am trying to write an action to setup flutter toolchain, but I can't extract zip.

The repo is https://github.com/chunlea/actions-setup-flutter, and code is here: https://github.com/chunlea/actions-setup-flutter/blob/master/src/installer.ts#L144

and I got this error:

```
2019-08-25T03:38:14.4067220Z ##[group]Run chunlea/actions-setup-flutter@releases/vtest
2019-08-25T03:38:14.4067400Z with:
2019-08-25T03:38:14.4067450Z   flutter-version: stable
2019-08-25T03:38:14.4067540Z ##[endgroup]
2019-08-25T03:38:15.1769030Z the file name is flutter_macos_v1.7.8+hotfix.4-stable.zip and .zip
2019-08-25T03:38:35.8228380Z [command]/Users/runner/runners/2.157.0/work/_actions/chunlea/actions-setup-flutter/releases/vtest/node_modules/@actions/tool-cache/scripts/externals/unzip-darwin /Users/runner/runners/2.157.0/work/_temp/a64b2350-0bc0-439f-99e4-e9eb03603b23
2019-08-25T03:38:35.8348120Z ##[error]There was an error when attempting to execute the process '/Users/runner/runners/2.157.0/work/_actions/chunlea/actions-setup-flutter/releases/vtest/node_modules/@actions/tool-cache/scripts/externals/unzip-darwin'. This may indicate the process failed to start. Error: spawn /Users/runner/runners/2.157.0/work/_actions/chunlea/actions-setup-flutter/releases/vtest/node_modules/@actions/tool-cache/scripts/externals/unzip-darwin ENOENT
2019-08-25T03:38:35.8387530Z ##[error]Node run failed with exit code 1
2019-08-25T03:38:35.8504660Z Cleaning up orphan processes
```

But if I change the code in `@actions/tool-cache`, not using the `unzip` in `node_modules`, using the `unzip` in system, it works.

any idea?

--- a/packages/tool-cache/__tests__/tool-cache.test.ts
+++ b/packages/tool-cache/__tests__/tool-cache.test.ts
@@ -142,6 +142,36 @@ describe('@actions/tool-cache', function() {
       }
     })
 
+    it('extracts a 7z to a directory that does not exist', async () => {
+      const tempDir = path.join(__dirname, 'test-install-7z')
+      const destDir = path.join(tempDir, 'not-exist')
+      try {
+        await io.mkdirP(tempDir)
+
+        // copy the 7z file to the test dir
+        const _7zFile: string = path.join(tempDir, 'test.7z')
+        await io.cp(path.join(__dirname, 'data', 'test.7z'), _7zFile)
+
+        // extract/cache
+        const extPath: string = await tc.extract7z(_7zFile, destDir)
+        await tc.cacheDir(extPath, 'my-7z-contents', '1.1.0')
+        const toolPath: string = tc.find('my-7z-contents', '1.1.0')
+
+        expect(extPath).toContain('not-exist')
+        expect(fs.existsSync(toolPath)).toBeTruthy()
+        expect(fs.existsSync(`${toolPath}.complete`)).toBeTruthy()
+        expect(fs.existsSync(path.join(toolPath, 'file.txt'))).toBeTruthy()
+        expect(
+          fs.existsSync(path.join(toolPath, 'file-with-ç-character.txt'))
+        ).toBeTruthy()
+        expect(
+          fs.existsSync(path.join(toolPath, 'folder', 'nested-file.txt'))
+        ).toBeTruthy()
+      } finally {
+        await io.rmRF(tempDir)
+      }
+    })
+
     it('extract 7z using custom 7z tool', async function() {
       const tempDir = path.join(
         __dirname,
@@ -212,6 +242,39 @@ describe('@actions/tool-cache', function() {
       await tc.cacheDir(extPath, 'my-tgz-contents', '1.1.0')
       const toolPath: string = tc.find('my-tgz-contents', '1.1.0')
 
+      expect(fs.existsSync(toolPath)).toBeTruthy()
+      expect(fs.existsSync(`${toolPath}.complete`)).toBeTruthy()
+      expect(fs.existsSync(path.join(toolPath, 'file.txt'))).toBeTruthy()
+      expect(
+        fs.existsSync(path.join(toolPath, 'file-with-ç-character.txt'))
+      ).toBeTruthy()
+      expect(
+        fs.existsSync(path.join(toolPath, 'folder', 'nested-file.txt'))
+      ).toBeTruthy()
+      expect(
+        fs.readFileSync(
+          path.join(toolPath, 'folder', 'nested-file.txt'),
+          'utf8'
+        )
+      ).toBe('folder/nested-file.txt contents')
+    })
+
+    it('extract .tar.gz to a directory that does not exist', async () => {
+      const tempDir = path.join(tempPath, 'test-install-tar.gz')
+      const destDir = path.join(tempDir, 'not-exist')
+
+      await io.mkdirP(tempDir)
+
+      // copy the .tar.gz file to the test dir
+      const _tgzFile: string = path.join(tempDir, 'test.tar.gz')
+      await io.cp(path.join(__dirname, 'data', 'test.tar.gz'), _tgzFile)
+
+      // extract/cache
+      const extPath: string = await tc.extractTar(_tgzFile, destDir)
+      await tc.cacheDir(extPath, 'my-tgz-contents', '1.1.0')
+      const toolPath: string = tc.find('my-tgz-contents', '1.1.0')
+
+      expect(extPath).toContain('not-exist')
       expect(fs.existsSync(toolPath)).toBeTruthy()
       expect(fs.existsSync(`${toolPath}.complete`)).toBeTruthy()
       expect(fs.existsSync(path.join(toolPath, 'file.txt'))).toBeTruthy()
@@ -358,6 +421,60 @@ describe('@actions/tool-cache', function() {
       } finally {
         await io.rmRF(destDir)
       }
+    } finally {
+      await io.rmRF(tempDir)
+    }
+  })
+
+  it('extract zip to a directory that does not exist', async function() {
+    const tempDir = path.join(__dirname, 'test-install-zip')
+    try {
+      await io.mkdirP(tempDir)
+
+      // stage the layout for a zip file:
+      //   file.txt
+      //   folder/nested-file.txt
+      const stagingDir = path.join(tempDir, 'zip-staging')
+      await io.mkdirP(path.join(stagingDir, 'folder'))
+      fs.writeFileSync(path.join(stagingDir, 'file.txt'), '')
+      fs.writeFileSync(path.join(stagingDir, 'folder', 'nested-file.txt'), '')
+
+      // create the zip
+      const zipFile = path.join(tempDir, 'test.zip')
+      await io.rmRF(zipFile)
+      if (IS_WINDOWS) {
+        const escapedStagingPath = stagingDir.replace(/'/g, "''") // double-up single quotes
+        const escapedZipFile = zipFile.replace(/'/g, "''")
+        const powershellPath = await io.which('powershell', true)
+        const args = [
+          '-NoLogo',
+          '-Sta',
+          '-NoProfile',
+          '-NonInteractive',
+          '-ExecutionPolicy',
+          'Unrestricted',
+          '-Command',
+          `$ErrorActionPreference = 'Stop' ; Add-Type -AssemblyName System.IO.Compression.FileSystem ; [System.IO.Compression.ZipFile]::CreateFromDirectory('${escapedStagingPath}', '${escapedZipFile}')`
+        ]
+        await exec.exec(`"${powershellPath}"`, args)
+      } else {
+        const zipPath: string = await io.which('zip')
+        await exec.exec(zipPath, [zipFile, '-r', '.'], {cwd: stagingDir})
+      }
+
+      const destDir = path.join(tempDir, 'not-exist')
+
+      const extPath: string = await tc.extractZip(zipFile, destDir)
+      await tc.cacheDir(extPath, 'foo', '1.1.0')
+      const toolPath: string = tc.find('foo', '1.1.0')
+
+      expect(extPath).toContain('not-exist')
+      expect(fs.existsSync(toolPath)).toBeTruthy()
+      expect(fs.existsSync(`${toolPath}.complete`)).toBeTruthy()
+      expect(fs.existsSync(path.join(toolPath, 'file.txt'))).toBeTruthy()
+      expect(
+        fs.existsSync(path.join(toolPath, 'folder', 'nested-file.txt'))
+      ).toBeTruthy()
     } finally {
       await io.rmRF(tempDir)
     }

--- a/packages/tool-cache/src/tool-cache.ts
+++ b/packages/tool-cache/src/tool-cache.ts
@@ -130,7 +130,7 @@ export async function extract7z(
   ok(IS_WINDOWS, 'extract7z() not supported on current OS')
   ok(file, 'parameter "file" is required')
 
-  dest = dest || (await _createExtractFolder(dest))
+  dest = await _createExtractFolder(dest)
 
   const originalCwd = process.cwd()
   process.chdir(dest)
@@ -199,7 +199,7 @@ export async function extractTar(
     throw new Error("parameter 'file' is required")
   }
 
-  dest = dest || (await _createExtractFolder(dest))
+  dest = await _createExtractFolder(dest)
   const tarPath: string = await io.which('tar', true)
   await exec(`"${tarPath}"`, [flags, '-C', dest, '-f', file])
 
@@ -218,7 +218,7 @@ export async function extractZip(file: string, dest?: string): Promise<string> {
     throw new Error("parameter 'file' is required")
   }
 
-  dest = dest || (await _createExtractFolder(dest))
+  dest = await _createExtractFolder(dest)
 
   if (IS_WINDOWS) {
     await extractZipWin(file, dest)

